voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/PermissionRequest.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2022 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.voice.impl
+
+import android.Manifest
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.provider.Settings
+import androidx.activity.result.ActivityResultCaller
+import androidx.core.app.ActivityCompat
+import com.duckduckgo.app.statistics.pixels.Pixel
+import com.duckduckgo.di.scopes.ActivityScope
+import com.duckduckgo.voice.impl.ActivityResultLauncherWrapper.Action.LaunchPermissionRequest
+import com.duckduckgo.voice.impl.ActivityResultLauncherWrapper.Request
+import com.duckduckgo.voice.store.VoiceSearchRepository
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+interface PermissionRequest {
+    fun registerResultsCallback(
+        caller: ActivityResultCaller,
+        activity: Activity,
+        onPermissionsGranted: () -> Unit
+    )
+
+    fun launch()
+}
+
+@ContributesBinding(ActivityScope::class)
+class MicrophonePermissionRequest @Inject constructor(
+    private val pixel: Pixel,
+    private val voiceSearchRepository: VoiceSearchRepository,
+    private val voiceSearchPermissionDialogsLauncher: VoiceSearchPermissionDialogsLauncher,
+    private val activityResultLauncherWrapper: ActivityResultLauncherWrapper
+) : PermissionRequest {
+    companion object {
+        private const val SCHEME_PACKAGE = "package"
+    }
+
+    private lateinit var _activity: Activity
```

Comment:
Instead of keeping reference to the `Activity`, would it not be better to pass the `Activity` as a parameter in the `launch()` method?

Reply:
Ahh good point. This is probably a result of me refactoring this class a lot of times. Will update!